pallets/parachain-staking/src/lib.rs

```diff
@@ -786,22 +949,44 @@ pub mod pallet {
 				Error::<T>::CandidateExists
 			);
 			T::Currency::reserve(&acc, bond)?;
-			let candidate = Collator::new(acc.clone(), bond);
 			let new_total = <Total<T>>::get() + bond;
 			<Total<T>>::put(new_total);
-			<CollatorState<T>>::insert(&acc, candidate);
+			count.increment::<T>();
+			<CandidateCount<T>>::put(count);
+			<CollatorState<T>>::insert(&acc, Collator::new(acc.clone(), bond));
 			<CandidatePool<T>>::put(candidates);
 			Self::deposit_event(Event::JoinedCollatorCandidates(acc, bond, new_total));
 			Ok(().into())
 		}
-		/// Request to leave the set of candidates. If successful, the account is immediately
+		/// Schedule exit from the set of candidates. If successful, the account is immediately
 		/// removed from the candidate pool to prevent selection as a collator, but unbonding is
-		/// executed with a delay of `BondDuration` rounds.
-		#[pallet::weight(0)]
-		pub fn leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+		/// executed after `BondDuration` rounds.
+		/// - `collator_count` weight hint is the number of collator candidates in CandidatePool
+		/// - `nominator_count` weight hint is the number of nominators for the collator leaving
+		#[pallet::weight(
+			T::WeightInfo::leave_candidates(*collator_count) +
+			T::WeightInfo::force_leave_candidates(
+				*collator_count,
+				*nominator_count,
+			)
```

Comment:
Caller pays for operations associated with collator exit when they make the call to trigger the exit.

Reply:
This just isn't the right way to do things. Instead, charge `T::WeightInfo::force_leave_candidates()` and then we should add the number of collator exits * `T::WeightInfo::force_leave_candidates()` to the `on_initialize` weight computation.

Reply:
It would be better for the caller to pay for the operation than to eat the cost in `on_initialize`, but we also want accurate weight estimates when including transactions in a block.